### PR TITLE
repass

### PR DESCRIPTION
**Pull request by quangtmbkhn**
1

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
   resources :users
   resources :account_activations, only: :edit
+  resources :password_resets, only: %i(new create edit update)
```
Comment (manhcuongdtbk) [inline config/routes.rb]: resources :password_resets, except: %i(index show destroy)

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user,   only: [:edit, :update]
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: before_action :get_user, only: %i(edit update)

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  before_action :valid_user, only: [:edit, :update]
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: change to %i

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  before_action :check_expiration, only: [:edit, :update]
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: change to %i

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  def new
+  end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = "Email sent with password reset instructions"
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: i18n

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      redirect_to root_url
+    else
+      flash.now[:danger] = "Email address not found"
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: i18n

File: app/controllers/password_resets_controller.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      render 'new'
+    end
+  end
+
+  def edit
+  end
+
+  def update
+    if params[:user][:password].empty?               
+      @user.errors.add(:password, "can't be empty")
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: i18n

File: app/controllers/password_resets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      render 'edit'
+    elsif @user.update_attributes(user_params)  
+      log_in @user
+      @user.update_attribute(:reset_digest, nil)
+      flash[:success] = "Password has been reset."
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: i18n

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  def valid_user
+    unless (@user && @user.activated? &&
+            @user.authenticated?(:reset, params[:id]))
+      redirect_to root_url
+    end
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
+      flash[:danger] = "Password reset has expired."
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: i18n

File: app/models/user.rb [hunk 10]
```diff
@@ -63,6 +63,20 @@ def new_token
     end
   end
 
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest,  User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
```
Comment (manhcuongdtbk) [inline app/models/user.rb]: remove ()

File: app/views/password_resets/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<% provide(:title, 'Reset password') %>
```
Comment (manhcuongdtbk) [inline app/views/password_resets/edit.html.erb]: i18n, remove ()

File: app/views/password_resets/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+      <%= hidden_field_tag :email, @user.email %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: 'form-control' %>
+
+      <%= f.submit "Update password", class: "btn btn-primary" %>
```
Comment (manhcuongdtbk) [inline app/views/password_resets/edit.html.erb]: i18n

File: app/views/password_resets/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<% provide(:title, "Forgot password") %>
```
Comment (manhcuongdtbk) [inline app/views/password_resets/new.html.erb]: i18n, remove ()

File: app/views/sessions/new.html.erb [hunk 14]
```diff
@@ -8,6 +8,7 @@
       <%= f.email_field :email, class: "form-control" %>
 
       <%= f.label :password %>
+      <%= link_to "(forgot password)", new_password_reset_path %>
```
Comment (manhcuongdtbk) [inline app/views/sessions/new.html.erb]: i18n

File: app/controllers/password_resets_controller.rb [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    end
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
+      flash[:danger] = ".flash_danger"  
+      redirect_to new_password_reset_url
+    end
+  end
+
```
Comment (chintk) [inline app/controllers/password_resets_controller.rb]: remove line

File: app/assets/javascripts/password_resets.coffee [hunk 16]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (manhcuongdtbk) [inline app/assets/javascripts/password_resets.coffee]: remove unused file

File: app/assets/stylesheets/password_resets.scss [hunk 17]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (manhcuongdtbk) [inline app/assets/stylesheets/password_resets.scss]: remove unused file

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = t ".flash_info"                     
+      redirect_to root_url
+    else
+      flash.now[:danger] = t ".flash_danger"  
+      render 'new'
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: render :new

File: app/controllers/password_resets_controller.rb [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    end
+  end
+
+  def edit
+  end
+
+  def update
+    if params[:user][:password].empty?               
+      @user.errors.add :password, t(".cant_be_empty")
+      render 'edit'
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: render :edit

File: app/controllers/password_resets_controller.rb [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+
+  def get_user
+    @user = User.find_by(email: params[:email])
```
Comment (manhcuongdtbk) [inline app/controllers/password_resets_controller.rb]: remove ()

**Pull request closed without merging**